source/s3/s3.go

```diff
@@ -0,0 +1,81 @@
+package s3
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/MiLk/nsscache-go/cache"
+	"github.com/MiLk/nsscache-go/source"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/pkg/errors"
+)
+
+type S3Source struct {
+	prefix string
+	bucket string
+	client s3iface.S3API
+}
+
+type JSONArray []map[string]interface{}
+
+func CreateS3Source(client s3iface.S3API, prefix string, bucket string) source.Source {
+	return &S3Source{
+		client: client,
+		prefix: prefix,
+		bucket: bucket,
+	}
+}
+
+func (s *S3Source) run(key string, c *cache.Cache, createEntry func() cache.Entry) error {
+	if s.prefix != "" {
+		key = fmt.Sprintf("%s/%s", s.prefix, key)
+	}
+
+	data, err := DownloadS3Data(s.client, s.bucket, key)
+
+	if err != nil {
+		return errors.Wrap(err, "downloading from S3")
+	}
+
+	r := make(JSONArray, 0)
+	err = json.Unmarshal([]byte(data), &r)
+
+	if err != nil {
+		return errors.Wrap(err, "json decoding")
+	}
+
+	for _, elem := range r {
+		str, _ := json.Marshal(elem)
+
+		e := createEntry()
+		err = json.Unmarshal([]byte(str), e)
```

Comment:
You can handle the error in a new local scope. See previous comment.

Reply:
I get an error trying to do that here:

![image](https://user-images.githubusercontent.com/7250722/31813341-47e51818-b586-11e7-821f-02737f16af52.png)

Reply:
Ok, I was forgetting the `if` ¬¬